### PR TITLE
Feature Worker Pools

### PR DESCRIPTION
**Pull request by mosquito**

### CODE REVIEW

File: aiomisc/worker_pool.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    REQUEST = 0
+    EXCEPTION = 1
+    RESULT = 2
+
+
+INET_AF = socket.AF_INET6
+
+
+def _inner(address: AddressType, cookie: bytes) -> None:
+    def step() -> Optional[bool]:
```
Comment (Pavkazzz) [inline aiomisc/worker_pool.py:39]: just bool is ok here

File: aiomisc/worker_pool.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+                    log.debug(
+                        "Process %r[%d] dead with exitcode %r, respawning",
+                        process, process.pid, process.exitcode
+                    )
+                    self.processes.remove(process)
+                    task = self.loop.create_task(self._respawn_process())
+                    self.task_store.add(task)
+                    task.add_done_callback(self.task_store.remove)
+
+            await asyncio.sleep(0.1)
```
Comment (Pavkazzz) [inline aiomisc/worker_pool.py:263]: Mb give ability to configure sleep time?

**Review by Pavkazzz — CHANGES_REQUESTED** [review]

File: aiomisc/worker_pool/pool.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            payload = await reader.readexactly(payload_length)
+            data = pickle.loads(payload)
+            return PacketTypes(packet_type), data
+
+        async def send(packet_type: PacketTypes, data: Any) -> None:
+            payload = pickle.dumps(data)
+            header = Header.pack(packet_type.value, len(payload))
+            writer.write(header)
+            writer.write(payload)
+            await writer.drain()
```
Comment (dizballanze) [inline aiomisc/worker_pool/pool.py:118]: why?

File: aiomisc/worker_pool/pool.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    @threaded
+    def __create_process(self, identity: str) -> Popen:
+        if self.__closing:
+            raise RuntimeError("Pool closed")
+
+        process = Popen(
+            [sys.executable, "-m", "aiomisc.worker_pool.process"],
+            stdin=PIPE, env=os.environ
+        )
```
Comment (dizballanze) [inline aiomisc/worker_pool/pool.py:63]: I think debug log with process id can be useful here

File: aiomisc/worker_pool/pool.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+                address="localhost",
+                port=0,
+            )
+            self.address = self.socket.getsockname()[:2]
+
+    @staticmethod
+    def _kill_process(process: Popen) -> None:
+        if process.returncode is not None:
+            return None
+        process.kill()
```
Comment (dizballanze) [inline aiomisc/worker_pool/pool.py:53]: Debug log here

File: aiomisc/worker_pool/process.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
+import logging
+import pickle
+import socket
+import sys
+from os import urandom
+from typing import Tuple, Union, Any
```
Comment (dizballanze) [inline aiomisc/worker_pool/process.py:6]: empty line missed

File: aiomisc/worker_pool/process.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    identity: str
+
+    address, cookie, identity = pickle.load(sys.stdin.buffer)
+
+    family = (
+        socket.AF_UNIX if isinstance(address, str) else INET_AF
+    )
+
+    with socket.socket(family, socket.SOCK_STREAM) as sock:
+        logging.debug("Connecting...")
```
Comment (dizballanze) [inline aiomisc/worker_pool/process.py:24]: add pid here and probably in all logs in this module

File: docs/source/locale/ru/LC_MESSAGES/worker_pool.po [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+msgid "Worker Pool"
+msgstr "Worker Pool"
+
+#: ../../source/worker_pool.rst:4
+msgid ""
+"Python has multiprocessing module with Pool class which implements "
+"multiprocess worker pool. The IPC in this case use completely synchronous"
+" communication method. This module reimplement the process based worker "
+"pool but IPC is completely asynchronous on callee side, but workers in "
+"separate processes no asynchronous."
```
Comment (dizballanze) [inline docs/source/locale/ru/LC_MESSAGES/worker_pool.po:16]: aren't asynchronous

File: docs/source/locale/ru/LC_MESSAGES/worker_pool.po [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+"вызывающей стороне,воркер процессы при этом не асинхронны, так-же как и в"
+" multiprocessing."
+
+#: ../../source/worker_pool.rst:11
+msgid "Example"
+msgstr "Пример"
+
+#: ../../source/worker_pool.rst:13
+msgid ""
+"This very be useful when you want to process data in separate process but"
```
Comment (dizballanze) [inline docs/source/locale/ru/LC_MESSAGES/worker_pool.po:31]: This will be useful

File: docs/source/locale/ru/LC_MESSAGES/worker_pool.po [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+" the input and result data are not big. For example parallel image "
+"processing. Lets write the program which accepts JPEG images and creates "
+"thumbnails for this. In this case you have the file with original image "
```
Comment (dizballanze) [inline docs/source/locale/ru/LC_MESSAGES/worker_pool.po:34]: for this -> for them

File: docs/source/locale/ru/LC_MESSAGES/worker_pool.po [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+msgstr ""
+"Это очень полезно, если вы хотите обрабатывать данные в отдельном "
+"процессе, но входные и результирующие данные невелики. Например "
+"параллельная обработка изображений. Давайте напишем программу, которая "
+"принимает изображения в формате JPEG и создает для этого эскизы. В этом "
+"случае у вас есть файл с исходным изображением, и вам нужно сгенерировать"
+" выходной путь для функции ``thumbnail``."
+
+#: ../../source/worker_pool.rst:21
+msgid "You have to install Pollow image processing library for run this example."
```
Comment (dizballanze) [inline docs/source/locale/ru/LC_MESSAGES/worker_pool.po:45]: Pollow -> Pillow
for run -> to run

File: tests/test_worker_pool.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+async def test_success(worker_pool):
+    results = await asyncio.gather(*[
+        worker_pool.create_task(operator.mul, i, i)
+        for i in range(worker_pool.workers * 2)
+    ])
+
+    results = sorted(results)
+
+    assert results == [i * i for i in range(worker_pool.workers * 2)]
```
Comment (dizballanze) [inline tests/test_worker_pool.py:26]: `assert sorted(results) == ...` no need for variable change

**Review by dizballanze — CHANGES_REQUESTED** [review]

**Review by dizballanze — APPROVED** [review]

**Review by alvassin — CHANGES_REQUESTED** [review]

**Review by Pavkazzz — APPROVED** [review]

**Pull request merged**